session.go

```diff
@@ -182,29 +182,32 @@ func (s *sessionCtx) id() string {
 	return string(s.session.ID)
 }
 
-func (s *sessionCtx) run() {
+func (s *sessionCtx) run(deadline time.Time) {
 	timeout := s.session.TTL / 3
 	ticker := time.NewTicker(timeout)
 	defer ticker.Stop()
 
 	for {
 		select {
-		case <-ticker.C:
 		case <-s.done:
 			return
 		case <-s.ctx.Done():
 			s.cancelWithError(s.ctx.Err())
 			return
-		}
-
-		// TODO: tolerate failures here? add retries with backoff?
-		renewSessionCtx, renewSessionCancel := context.WithTimeout(s, timeout)
-		err := s.session.Client.renewSession(renewSessionCtx, s.id())
-		renewSessionCancel()
-
-		if err != nil {
-			s.cancelWithError(err)
-			return
+		case now := <-ticker.C:
+			renewSessionCtx, renewSessionCancel := context.WithTimeout(s, timeout)
+			err := s.session.Client.renewSession(renewSessionCtx, s.id())
+			renewSessionCancel()
+
+			if err != nil {
+				if now.Before(deadline) {
+					continue
```

Comment:
Should we log it anyway?  Could see this potentially causing issues if we get caught in a fail loop.

Reply:
If it fails enough times it'll reach the TTL eventually. I'd rather not add logging into this package directly because it has none currently (plus there's still no good way of abstracting logging in the standard library).
The package allows the http transport to be configured, so there's opportunity there to inject logging and metrics that would surface errors or non-successful http status codes.

Reply:
Figured as much.  Hopefully this doesn't cause a stampeding herd type situation.